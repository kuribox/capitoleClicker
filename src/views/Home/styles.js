import { StyleSheet } from 'react-native';
import variables from '../../theme/variables';

export default StyleSheet.create({
  scrollContainer: {
    backgroundColor: variables.backgroundColor,
    height: 200,
    flexDirection: 'column',
    flexGrow: 1,
  },
  container: {
    alignItems: 'center',
    justifyContent: 'space-between',
    backgroundColor: variables.backgroundColor,
    padding: variables.padding
  },
  loginContainer: {
    display: "flex",
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    width: "100%",
    paddingVertical: variables.padding,
    marginBottom: 20,
  },
  logo: {
    maxWidth: "80%",
    resizeMode: 'contain'
  },
  label: {
    fontWeight: "bold",
    fontSize: 20,
  },
  input: {
    borderColor: variables.primary,
    borderRadius: 5,
    height: 40,
    width: '100%',
    margin: 12,
    borderWidth: 1,
    padding: 10,
  },
  userListTitle: {
    fontWeight: "bold",
    fontSize: 18,
  },
  userListContainer: {
    flex: 1,
    width: "100%",
  },
  userListElement: {
    paddingVertical: variables.padding,
    borderBottomColor: variables.primary,
    borderBottomWidth: 1,
    flexDirection: "row",
    justifyContent: "space-between"
  }
});
