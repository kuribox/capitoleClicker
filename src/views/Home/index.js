import React, { useState, useEffect, useCallback } from 'react';
import { TextInput, Text, ScrollView, Image, View, Button, TouchableOpacity } from 'react-native';
import { useStore } from '../../providers/StoreContext';
import LogoImg from '../../public/logo.png';
import styles from './styles';
import variables from '../../theme/variables';
import { useFocusEffect } from '@react-navigation/native';

const Home = ({navigation}) => {
  const { localStorage, setUser } = useStore();
  const [user, setUserValue] = useState(null);
  const [userList, setUserList] = useState(null);

  useFocusEffect(
    useCallback(() => {
      localStorage.load({ key: 'users'}).then((e) => {
        setUserList(e);
      }).catch(() => {
        localStorage.save({ key: 'users', data: [] })
        setUserList([]);
      })
    }, [])
  );

  useEffect(() => {
    if (!userList) return;
    localStorage.load({ key: 'users'}).then((e) => {
      if (JSON.stringify(e) !== JSON.stringify(userList)) {
        localStorage.save({ key: 'users', data: userList })
      }
    }).catch(e => console.error(e))
  }, [userList]);

  const login = (currentUser) => {
    let userData = currentUser;
    if (!userData) {
      userData = { name: user, value: 0, clickers: []};
      localStorage.load({ key: 'users'}).then((e) => {
        setUserList([...e, userData]);
        setUser(userData);
        navigation.navigate('Clicker');
      })
    } else {
      setUser(userData);
      navigation.navigate('Clicker');
    }
  }

  return (
    <ScrollView style={styles.scrollContainer} contentContainerStyle={styles.container}>
      <Image style={styles.logo} source={LogoImg} />

      <View style={styles.loginContainer}>
        <Text style={styles.label}>
          Please enter your username
        </Text>
      
        <TextInput
          testID="user-input"
          style={styles.input}
          value={user}
          onChangeText={(v) => {
            setUserValue(v);
          }}
        />
        <Button
          testID="login-button"
          color={variables.primary}
          disabled={!user || user.lenght < 3}
          title="Join"
          onPress={() => {
            login(userList.find(e => e.name === user))
          }}
        />
      </View>

      <View style={styles.userListContainer}>
        {userList && userList.length > 0 && <Text style={styles.userListTitle}>Top 5 scores</Text>}
        {userList && userList.sort((a, b) => a.value > b.value ? -1 : 1 ).slice(0, 5).map((e) => (
          <TouchableOpacity onPress={() => login(e)} key={e.name} style={styles.userListElement}>
            <Text>
              {e.name}
            </Text>
            <Text>
              {e.value} 
            </Text>
          </TouchableOpacity>
        ))}
      </View>
    </ScrollView>
  );
}

export default Home;

