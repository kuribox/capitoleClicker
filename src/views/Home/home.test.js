import React from 'react';
import renderer from 'react-test-renderer';
import Home from './index';
import TestContainer, { Navigation } from '../../../test/TestContainer';
import {render, waitFor, screen, fireEvent} from '@testing-library/react-native'
import AsyncStorage from '@react-native-async-storage/async-storage';

it('renders correctly', () => {
  const testRenderer = renderer.create(<TestContainer Component={Home} />);
  expect(testRenderer.toJSON()).toMatchSnapshot();
});

it('Disabled button', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={Home} />));   
  const button = queryByTestId('login-button');;
  expect(button.props.accessibilityState.disabled).toBe(true);
});

it('Empty input', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={Home} />));   
  const userInput = queryByTestId('user-input');  
  expect(userInput.props.value).toBeNull();
});

it('Edit input enable button', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={Home} />));   
  const userInput = queryByTestId('user-input');  
  const button = queryByTestId('login-button');;

  expect(button.props.accessibilityState.disabled).toBe(true);

  expect(userInput.props.value).toBeNull();

  fireEvent.changeText(userInput, 'name');
  expect(userInput.props.value).toBe('name');
  expect(button.props.accessibilityState.disabled).toBeUndefined();
});

it('Check login storage/navigate', async () => {
  const { queryByTestId } = await waitFor(() => render(<TestContainer Component={Home} />));   
  const userInput = queryByTestId('user-input');  
  const button = queryByTestId('login-button');;

  fireEvent.changeText(userInput, 'name');

  fireEvent.press(button)

  const root = await waitFor(() => {
    expect(Navigation.navigate).toBeCalledWith("Clicker");
  });

  expect(JSON.stringify(JSON.parse((await AsyncStorage.getItem("users"))).rawData)).toBe("[{\"name\":\"name\",\"value\":0,\"clickers\":[]}]")
});