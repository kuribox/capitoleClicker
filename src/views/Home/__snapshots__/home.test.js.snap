// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "justifyContent": "space-between",
      "padding": 10,
    }
  }
  style={
    Object {
      "backgroundColor": "white",
      "flexDirection": "column",
      "flexGrow": 1,
      "height": 200,
    }
  }
>
  <View>
    <Image
      source={1}
      style={
        Object {
          "maxWidth": "80%",
          "resizeMode": "contain",
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flex": 1,
          "justifyContent": "center",
          "marginBottom": 20,
          "paddingVertical": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Please enter your username
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "#072146",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 40,
            "margin": 12,
            "padding": 10,
            "width": "100%",
          }
        }
        testID="user-input"
        underlineColorAndroid="transparent"
        value={null}
      />
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="login-button"
      >
        <View
          style={
            Array [
              Object {},
              Object {},
            ]
          }
        >
          <Text
            disabled={true}
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#072146",
                },
                Object {
                  "color": "#cdcdcd",
                },
              ]
            }
          >
            Join
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "width": "100%",
        }
      }
    />
  </View>
</RCTScrollView>
`;
